### PR TITLE
 stats: convert small range to points in selectivity

### PR DESCRIPTION
**Pull request by lamxTyler**
### What problem does this PR solve? 

Previously, we always use histogram to estimate the range query, but if the range is small and the bucket is large, it is very likely that the estimation would be wrong.

### What is changed and how it works?

Convert small range to points and use cm sketch to estimate them.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - None

Related changes

 - None

### CODE REVIEW

File: statistics/table.go [hunk 1]
```diff
@@ -361,11 +361,34 @@ func isSingleColIdxNullRange(idx *Index, ran *ranger.Range) bool {
 	return false
 }
 
+func (coll *HistColl) getEqualCondSelectivity(idx *Index, bytes []byte, coverAll bool) float64 {
```
Comment (winoros) [inline statistics/table.go]: Add some comments for `coverAll`

File: statistics/scalar_test.go [hunk 2]
```diff
@@ -168,3 +176,69 @@ func (s *testStatisticsSuite) TestCalcFraction(c *C) {
...
+func (s *testStatisticsSuite) TestEnumRangeValues(c *C) {
+	tests := []struct {
+		low         types.Datum
+		high        types.Datum
+		lowExclude  bool
+		highExclude bool
+		res         string
+	}{
+		{
+			low:         types.NewIntDatum(0),
```
Comment (winoros) [inline statistics/scalar_test.go]: Add a overflow case?

File: statistics/scalar.go [hunk 3]
```diff
@@ -175,3 +177,122 @@ func calcFraction4Datums(lower, upper, value *types.Datum) float64 {
...
+	}
+	if highExclude {
+		exclude++
+	}
+	switch low.Kind() {
+	case types.KindInt64:
+		// Overflow check.
+		lowVal, highVal := low.GetInt64(), high.GetInt64()
+		if lowVal < 0 && highVal > 0 {
+			if lowVal <= -maxNumStep || highVal >= maxNumStep {
```
Comment (qw4990) [inline statistics/scalar.go]: Why do we need to limit the range in `[-maxNumStep, maxNumStep]`?

Comment (lamxTyler) [reply statistics/scalar.go]: Overflow check, combined with the upper `if`, we can make sure that it is safe and does not miss valid ranges.

Comment (lzmhhh123) [reply statistics/scalar.go]: I think these two `if` are preventing overflow `int64`.

Comment (qw4990) [reply statistics/scalar.go]: got

File: statistics/histogram.go [hunk 4]
```diff
@@ -742,6 +742,18 @@ func (c *Column) GetColumnRowCount(sc *stmtctx.StatementContext, ranges []*range
 			}
 			continue
 		}
+		rangeVals := enumRangeValues(rg.LowVal[0], rg.HighVal[0], rg.LowExclude, rg.HighExclude)
+		// The small range case.
```
Comment (qw4990) [inline statistics/histogram.go]: Does it seem that we don't check the condition `the bucket is large` now?
Or is there no necessary to check it?

Comment (lamxTyler) [reply statistics/histogram.go]: Which check do you mean?

File: statistics/scalar.go [hunk 5]
```diff
@@ -175,3 +177,122 @@ func calcFraction4Datums(lower, upper, value *types.Datum) float64 {
 	}
 	return 0.5
 }
+
+const maxNumStep = 10
+
+func enumRangeValues(low, high types.Datum, lowExclude, highExclude bool) []types.Datum {
+	if low.Kind() != high.Kind() {
```
Comment (eurekaka) [inline statistics/scalar.go]: When would this happen? like `(1, +inf]`?

File: statistics/scalar.go [hunk 6]
```diff
@@ -175,3 +177,122 @@ func calcFraction4Datums(lower, upper, value *types.Datum) float64 {
 	}
 	return 0.5
 }
+
+const maxNumStep = 10
```
Comment (eurekaka) [inline statistics/scalar.go]: Could we make this a variable correlated with average bucket width for example?

File: statistics/scalar.go (hunk 5 above)
Comment (lamxTyler) [reply statistics/scalar.go]: Yes.

File: statistics/scalar.go (hunk 6 above)
Comment (lamxTyler) [reply statistics/scalar.go]: But we still need a `max` step to limit the optimization time.

**Pull request merged**